synapse/storage/registration.py

```diff
@@ -984,3 +983,191 @@ def _bg_user_threepids_grandfather_txn(txn):
         yield self._end_background_update("user_threepids_grandfather")
 
         defer.returnValue(1)
+
+    @defer.inlineCallbacks
+    def get_threepid_validation_session(self, medium, address, client_secret):
+        """Gets the latest session_id and send_attempt (if available) for a
+        client_secret/medium/address combo
+
+        Args:
+            medium (str): The medium of the 3PID
+            address (str): The address of the 3PID
+            client_secret (str): A unique string provided by the client to
+                help identify this validation attempt
+
+        Returns:
+            deferred {str, int}|None: A dict containing the
+                latest session_id and send_attempt count for this 3PID.
+                Otherwise None if there hasn't been a previous attempt
+        """
+        row = yield self._simple_select_one(
+            table="threepid_validation_session",
+            keyvalues={
+                "medium": medium,
+                "address": address,
+                "client_secret": client_secret,
+            },
+            retcols=["session_id", "last_send_attempt"],
+            allow_none=True,
+            desc="get_threepid_validation_session",
+        )
+
+        defer.returnValue(row)
+
+    @defer.inlineCallbacks
+    def validate_threepid_validation_token(
+        self,
+        session_id,
+        client_secret,
+        token,
+        current_ts,
+    ):
+        """Validate a validation token, session_id and client_secret
+
+        Args:
+            session_id (str): The id of a validation session
+            client_secret (str): A unique string provided by the client to
+                help identify this validation attempt
+            token (str): A validation token
+            current_ts (int): The current unix time in milliseconds. Used for
+                checking token expiry status
+
+        Returns:
+            deferred (bool, str|None): A tuple where the first value represents
+                whether the provided sid/client_secret/token combo was correct,
+                and the second being the link to redirect the user to if there
+                is one
+        """
+        # Check that everything matches in the DB
+        def validate_threepid_validation_token_txn(
+            txn,
+            token,
+            client_secret,
+            ts,
+        ):
+            sql = ("SELECT token, next_link FROM threepid_validation_token "
+                   "WHERE token = ? AND session_id = (SELECT session_id FROM "
+                   "threepid_validation_session WHERE client_secret = ?) AND "
+                   "expires > ? LIMIT 1")
```

Comment:
This can be written as:

```python
sql = """
    SELECT token, next_link FROM threepid_validation_token
    INNER JOIN threepid_validation_session USING (session_id)
    WHERE token = ? AND client_secret = ? AND expires > ?
    LIMIT 1
"""
```